src/core/lib/security/credentials/external/pluggable_auth_external_account_credentials.cc

```diff
@@ -0,0 +1,401 @@
+//
+// Copyright 2023 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+#include <grpc/support/port_platform.h>
+
+#include "src/core/lib/security/credentials/external/pluggable_auth_external_account_credentials.h"
+
+#include <cxxabi.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <algorithm>
+#include <chrono>
+#include <future>
+#include <initializer_list>
+#include <map>
+#include <memory>
+#include <thread>
+#include <utility>
+
+#include "absl/status/status.h"
+#include "absl/status/statusor.h"
+#include "absl/strings/numbers.h"
+#include "absl/strings/str_format.h"
+#include "absl/strings/str_split.h"
+#include "absl/strings/string_view.h"
+#include "absl/strings/strip.h"
+
+#include <grpc/slice.h>
+#include <grpc/support/alloc.h>
+#include <grpc/support/json.h>
+#include <grpc/support/string_util.h>
+#include <grpc/support/time.h>
+
+#include "src/core/lib/gpr/subprocess.h"
+#include "src/core/lib/iomgr/load_file.h"
+#include "src/core/lib/json/json.h"
+#include "src/core/lib/json/json_reader.h"
+#include "src/core/lib/security/credentials/external/external_account_credentials.h"
+#include "src/core/lib/slice/slice.h"
+#include "src/core/lib/slice/slice_internal.h"
+
+extern char** environ;
+
+#define DEFAULT_EXECUTABLE_TIMEOUT_MS 30000  // 30 seconds
+#define MIN_EXECUTABLE_TIMEOUT_MS 5000       // 5 seconds
+#define MAX_EXECUTABLE_TIMEOUT_MS 120000     // 120 seconds
+#define SAML_SUBJECT_TOKEN_TYPE "urn:ietf:params:oauth:token-type:saml2"
+#define GOOGLE_EXTERNAL_ACCOUNT_ALLOW_EXECUTABLES \
+  "GOOGLE_EXTERNAL_ACCOUNT_ALLOW_EXECUTABLES"
+#define GOOGLE_EXTERNAL_ACCOUNT_ALLOW_EXECUTABLES_ACCEPTED_VALUE "1"
+
+inline bool isKeyPresent(absl::StatusOr<grpc_core::Json> json,
+                         std::string key) {
+  return json->object().find(key.c_str()) != json->object().end();
+}
+
+inline std::string getStringValue(absl::StatusOr<grpc_core::Json> json,
+                                  std::string key) {
+  auto it = json->object().find(key.c_str());
+  return it->second.string();
+}
+
+bool isExpired(int expiration_time) {
+  return gpr_time_cmp(
+             gpr_time_from_seconds(expiration_time, GPR_CLOCK_REALTIME),
+             gpr_now(GPR_CLOCK_REALTIME)) <= 0;
+}
```

Comment:
Maybe put in a central place?

Reply:
isExpired method is specific to pluggable auth as it uses Realtime clock and checks if time in "seconds" has expired, so is it fine to leave this here?
I've used an already existing method for isKeyPresent and getStringValue.